hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/stats/SidecarInstrumentationImpl.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.service.mono.stats;
+
+import com.hedera.services.stream.proto.SidecarType;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.EnumMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Holds the accumulators needed for the metrics/logging of a sidecar record (corresponding, most
+ * likely, to a record block. There's a separate accumulator for each kind of sidecar record
+ * (bytecode, state change, actions).
+ */
+public record SidecarInstrumentationImpl(
+        @NonNull EnumMap<SidecarType, SidecarAccumulators> accumulators,
+        @NonNull Clock clock,
+        @NonNull AtomicInteger copies)
+        implements SidecarInstrumentation {
+
+    /** Constructs the sidecar instrumentation */
+    public SidecarInstrumentationImpl() {
+        this(new EnumMap<>(SidecarType.class), Clock.systemUTC(), new AtomicInteger());
+        createAccumulators();
+    }
+
+    /** Constructs the sidecar instrumentation, allowing the injection of a `Clock` (for testing) */
+    public SidecarInstrumentationImpl(@NonNull Clock clock) {
+        this(new EnumMap<>(SidecarType.class), clock, new AtomicInteger());
+        createAccumulators();
+    }
+
+    /** Create an accumulator for each kind of sidecar record */
+    private void createAccumulators() {
+        for (final var e : SidecarType.values())
+            accumulators.put(e, new SidecarAccumulators(e.name()));
+    }
+
+    /**
+     * Add a sample of the serialized size of a sidecar transaction, for a particular kind of
+     * sidecar
+     */
+    @Override
+    public void addSample(@NonNull final SidecarType type, long size) {
+        final var acc = accumulators.get(type);
```

Comment:
Yes, I just read the new Java style policy and it's in there too - I'll fix that up.

Reply:
(Our new Java style guide suggests using `Objects.requireNonNull(object, message)` but gives no hint as to what the message should be.  It would be nice to have a suggestion.  Is the argument name by itself sufficient?

```
Objects.requireNonNull(type, "type");
```

maybe
```
Objects.requireNonNull(type, "type: SidecarType");
```

Or what? (Inquiring minds want to know.)

Reply:
The platform team uses a method `CommonUtils.throwArgNull` that takes a parameter and its name. When the parameter is `null`, it constructs a full error message: "The supplied argument '%s' cannot be null!". I like this approach, because it keeps the check as short as possible while providing an understandable error message.

Once the repositories are merged, we should agree on a single approach.